### Exercises I
Run 
```python
python -m pip install mysqlclient pymysql
```

from your terminal to install pymysql and the mysqlclient.

Create a notebook or python script named advanced_dataframes to do your work in for these exercises.

In [1]:
import pandas as pd
import numpy as np
np.random.seed(123)
from pydataset import data

1. Run `python -m pip install mysqlclient pymysql` from your terminal to install the mysql client (any folder is fine)

2. cd into your exercises folder for this module and run `echo env.py >> .gitignore`

3. Create a function named `get_db_url`. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [2]:
from env import host, user, password

get_db_url = f'mysql+pymysql://{user}:{password}@{host}/employees'

In [3]:
def get_db_url(db):
    return f'mysql+pymysql://{user}:{password}@{host}/{db}'

In [140]:
def sql(tbl):
    return f'SELECT * FROM {tbl}'

4. Use your function to obtain a connection to the `employees` database.

In [5]:
employees_db = pd.read_sql(sql('employees'), get_db_url('employees'))
employees_db.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

5. Once you have successfully run a query:
    - Intentionally make a typo in the database url. What kind of error message do you see?
        * ANSWER: OperationalError: (1044, "Access denied for user 'mirzakhani_1942'@'%' to database 'Boss'")
    - Intentionally make an error in your SQL query. What does the error message look like?
        * ANSWER: ProgrammingError: (1146, "Table 'employees.emplopers' doesn't exist")

6. Read the `employees` and `titles` tables into two separate DataFrames.

In [6]:
employees_df = pd.read_sql(sql('employees'), get_db_url('employees'))
employees_df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [7]:
titles_df = pd.read_sql(sql('titles'), get_db_url('employees'))
titles_df

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

7. How many rows and columns do you have in each DataFrame? Is that what you expected?
    
    * ANSWER: YES
              employees_df - 300024 rows × 6 columns
              titles_df - 443308 rows × 4 columns

In [8]:
employees_df.shape

(300024, 6)

In [9]:
titles_df.shape

(443308, 4)

8. Display the summary statistics for each DataFrame.

In [10]:
employees_df.info

<bound method DataFrame.info of         emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]>

In [11]:
employees_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [12]:
titles_df.info

<bound method DataFrame.info of         emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]>

In [13]:
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

9. How many unique titles are in the `titles` DataFrame?

In [14]:
print(titles_df.title.nunique())
list(titles_df.title.unique())

7


['Senior Engineer',
 'Staff',
 'Engineer',
 'Senior Staff',
 'Assistant Engineer',
 'Technique Leader',
 'Manager']

### Why doesn't this work?

In [16]:
# pd.read_sql('SELECT COUNT(DISTINCT title) AS unique_title_count FROM titles', get_db_url)
# testing direct call method

10. What is the oldest date in the `to_date` column? 

In [17]:
titles_df.to_date.min()

datetime.date(1985, 3, 1)

In [18]:
pd.read_sql(sql('titles'), get_db_url('employees')).to_date.min()
# testing direct call method

datetime.date(1985, 3, 1)

11. What is the most recent date in the `to_date` column?

### This works in SQL?
SELECT MAX(to_date) FROM titles WHERE to_date NOT LIKE '999%'

In [134]:
qu = "SELECT MAX(to_date) FROM titles WHERE to_date < 9999-01-01"
pd.read_sql(qu, get_db_url('employees'))

MAX(to_date)
0         None

### Exercises II

1. Copy the users and roles DataFrames from the examples above.

In [21]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [22]:
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

2. What is the result of using a right join on the DataFrames?

    * ANSWER: All data is kept from the table on the right ("roles" as the second table called with ``` ".merge(roles,"```) all missing data that does not match from the left "users" table will be filled in as "NaN".

In [23]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

3. What is the result of using an outer join on the DataFrames?

    * ANSWER: All records from both tables are kept with unmatched data filled as "NaN".

In [24]:
users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

    * ANSWER: It will error out if you leave the left_on and right_on as empty quotes ""
              It will mismatch your data if align them both on "id" without using the foriegn key
              For left and right with no "_on" statement, 
                  it will bring back headers for both,
                  it will fill data for only left or right (according to "how"),
                  no matches will be made between the two,
                  and all data from the column not in the primary table being kept (how= L or R)
                  will be filled with "NaN"
              Outer will return both full data tables 
                  with no matched data and all headers not in original table filled with "NaN"
              Inner will bring back only the headers from both tables because there ae no matched records

In [25]:
users.merge(roles, left_on='id', right_on='id', how='outer', indicator=True)

id name_x  role_id     name_y     _merge
0   1    bob      1.0      admin       both
1   2    joe      2.0     author       both
2   3  sally      3.0   reviewer       both
3   4   adam      3.0  commenter       both
4   5   jane      NaN        NaN  left_only
5   6   mike      NaN        NaN  left_only

In [26]:
users.merge(roles, how='left', indicator=True)

id   name  role_id     _merge
0   1    bob      1.0  left_only
1   2    joe      2.0  left_only
2   3  sally      3.0  left_only
3   4   adam      3.0  left_only
4   5   jane      NaN  left_only
5   6   mike      NaN  left_only

In [27]:
users.merge(roles, how='right', indicator=True)

id       name  role_id      _merge
0   1      admin      NaN  right_only
1   2     author      NaN  right_only
2   3   reviewer      NaN  right_only
3   4  commenter      NaN  right_only

In [28]:
users.merge(roles, how='outer', indicator=True)

id       name  role_id      _merge
0   1        bob      1.0   left_only
1   2        joe      2.0   left_only
2   3      sally      3.0   left_only
3   4       adam      3.0   left_only
4   5       jane      NaN   left_only
5   6       mike      NaN   left_only
6   1      admin      NaN  right_only
7   2     author      NaN  right_only
8   3   reviewer      NaN  right_only
9   4  commenter      NaN  right_only

In [29]:
users.merge(roles, how='inner', indicator=True)

Empty DataFrame
Columns: [id, name, role_id, _merge]
Index: []

5. Load the mpg dataset from PyDataset.

In [30]:
from pydataset import data
mpg = data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

6. Output and read the documentation for the mpg dataset.

In [31]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




7. How many rows and columns are in the dataset?

In [32]:
mpg.shape

(234, 11)

8. Check out your column names and perform any cleanup you may want on them.

In [33]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [34]:
mpg.columns

Index(['manufacturer', 'model', 'displ', 'year', 'cyl', 'trans', 'drv', 'cty',
       'hwy', 'fl', 'class'],
      dtype='object')

In [35]:
mpg['class'].unique()

array(['compact', 'midsize', 'suv', '2seater', 'minivan', 'pickup',
       'subcompact'], dtype=object)

In [36]:
type(mpg)

pandas.core.frame.DataFrame

In [37]:
mpg.rename(columns = {'class':'classification'}) # class is a reserved word in python

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

    classification  
1          compact  
2          compact  
3          compact  
4          compact  
5          compact  
..             ...  
230        midsize  
231        midsize  
232        midsize  
233        midsize  
234        midsize  

[234 rows x 11 columns]

9. Display the summary statistics for the dataset.

In [38]:
mpg.info

<bound method DataFrame.info of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    co

In [39]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

10. How many different manufacturers are there?

In [40]:
print(mpg['manufacturer'].nunique())
list(mpg['manufacturer'].unique())

15


['audi',
 'chevrolet',
 'dodge',
 'ford',
 'honda',
 'hyundai',
 'jeep',
 'land rover',
 'lincoln',
 'mercury',
 'nissan',
 'pontiac',
 'subaru',
 'toyota',
 'volkswagen']

11. How many different models are there?

In [41]:
print(mpg['model'].nunique())
list(mpg['model'].unique())

38


['a4',
 'a4 quattro',
 'a6 quattro',
 'c1500 suburban 2wd',
 'corvette',
 'k1500 tahoe 4wd',
 'malibu',
 'caravan 2wd',
 'dakota pickup 4wd',
 'durango 4wd',
 'ram 1500 pickup 4wd',
 'expedition 2wd',
 'explorer 4wd',
 'f150 pickup 4wd',
 'mustang',
 'civic',
 'sonata',
 'tiburon',
 'grand cherokee 4wd',
 'range rover',
 'navigator 2wd',
 'mountaineer 4wd',
 'altima',
 'maxima',
 'pathfinder 4wd',
 'grand prix',
 'forester awd',
 'impreza awd',
 '4runner 4wd',
 'camry',
 'camry solara',
 'corolla',
 'land cruiser wagon 4wd',
 'toyota tacoma 4wd',
 'gti',
 'jetta',
 'new beetle',
 'passat']

12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [42]:
mpg['mileage_difference'] = mpg['hwy'] - mpg['cty']
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

NOTE: There are different knds of "mean" (ht, wt, harmonic using rates, etc...)

https://machinelearningmastery.com/arithmetic-geometric-and-harmonic-means-for-machine-learning/

In [43]:
# Harmonic Mean
mpg['average_mileage'] = round(2 / ((1/mpg['hwy']) + (1/mpg['cty'])), 2)
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  average_mileage  
1    compact                  11            22.21  
2    compact                   8            24.36  
3    compact                  11            24.31  
4    compact                   9            24.71  
5    compact                  10            19.81  
..       ...                 ...              ...  
230  midsize                   9            22.64  
231  midsize                   8            24.36  
232  midsize                  10            19.81  
233  midsize                   8            21.27  
234  midsize                   9            20.56  

[234 rows x 13 columns]

In [44]:
mpg['average_mileage'] = (mpg['hwy'] + mpg['cty']) / 2
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

In [45]:
mpg.trans.unique()

array(['auto(l5)', 'manual(m5)', 'manual(m6)', 'auto(av)', 'auto(s6)',
       'auto(l4)', 'auto(l3)', 'auto(l6)', 'auto(s5)', 'auto(s4)'],
      dtype=object)

In [46]:
mpg['is_automatic'] = mpg['trans'].str.startswith('auto')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  mileage_difference  average_mileage  is_automatic  
1    compact                  11             23.5          True  
2    compact                   8             25.0         False  
3    compact                  11             25.5         False  
4    compact                   9             25.5          True  
5    compact                  10             21.0          True  
..       ...                 ...              ...           ...  
230  midsize                   9             23.5          True  
231  midsize                   8             25.0         False  
232  midsize                  10             21.0          True  
233  midsize                   8             22.0         False  
234  midsize                   9             21.5          True  

[234 rows x 14 columns]

15. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

In [47]:
mpg.groupby('manufacturer').average_mileage.mean().sort_values(ascending = False)

manufacturer
honda         28.500000
volkswagen    25.074074
hyundai       22.750000
subaru        22.428571
audi          22.027778
toyota        21.720588
pontiac       21.700000
nissan        21.346154
chevrolet     18.447368
ford          16.680000
mercury       15.625000
jeep          15.562500
dodge         15.540541
lincoln       14.166667
land rover    14.000000
Name: average_mileage, dtype: float64

In [48]:
mpg.groupby('manufacturer').average_mileage.mean().nlargest(1)

manufacturer
honda    28.5
Name: average_mileage, dtype: float64

#### My original answer before exercise review

In [49]:
mpg['average_mileage'].max()

39.5

In [50]:
mpg[mpg.average_mileage == mpg.average_mileage.max()]

manufacturer       model  displ  year  cyl       trans drv  cty  hwy fl  \
222   volkswagen  new beetle    1.9  1999    4  manual(m5)   f   35   44  d   

          class  mileage_difference  average_mileage  is_automatic  
222  subcompact                   9             39.5         False

In [51]:
mpg.average_mileage.sort_values(ascending = False)

222    39.5
213    38.5
223    35.0
197    32.5
196    30.5
       ... 
127    10.5
60     10.5
66     10.5
70     10.5
55     10.5
Name: average_mileage, Length: 234, dtype: float64

16. Do automatic or manual cars have better miles per gallon?

In [52]:
mpg['trans_category'] = np.where(mpg.trans.str.startswith('auto'), 'automatic', 'manual')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class  \
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact   
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact   
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact   
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact   
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact   

   mileage_difference  average_mileage  is_automatic trans_category  
1                  11             23.5          True      automatic  
2                   8             25.0         False         manual  
3                  11             25.5         False         manual  
4                   9             25.5          True      automatic  
5                  10             21.0          True      automatic

In [53]:
mpg.groupby('trans_category')[['average_mileage']].mean().round(1)

average_mileage
trans_category                 
automatic                  19.1
manual                     22.2

### Exercises III

1. Use your get_db_url function to help you explore the data from the chipotle database.

In [115]:
chipotle_db = pd.read_sql(sql('orders'), get_db_url('chipotle'))
chipotle_db.head()

id  order_id  quantity                              item_name  \
0   1         1         1           Chips and Fresh Tomato Salsa   
1   2         1         1                                   Izze   
2   3         1         1                       Nantucket Nectar   
3   4         1         1  Chips and Tomatillo-Green Chili Salsa   
4   5         2         2                           Chicken Bowl   

                                                           choice_description  \
0                                                                         nan   
1                                                                [Clementine]   
2                                                                     [Apple]   
3                                                                         nan   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   

  item_price  
0     $2.39   
1     $3.39   
2     $3.39   
3     $2.39   
4    $16.98

2. What is the total price for each order?

In [116]:
chipotle_db.dtypes

id                     int64
order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

In [117]:
chipotle_db['item_price'] = chipotle_db['item_price'].str.replace('$', '').astype('float')

/var/folders/x8/x12zllv954v785j2tdhggqrr0000gn/T/ipykernel_4025/3754449823.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  chipotle_db['item_price'] = chipotle_db['item_price'].str.replace('$', '').astype('float')


In [118]:
chipotle_db.head()

id  order_id  quantity                              item_name  \
0   1         1         1           Chips and Fresh Tomato Salsa   
1   2         1         1                                   Izze   
2   3         1         1                       Nantucket Nectar   
3   4         1         1  Chips and Tomatillo-Green Chili Salsa   
4   5         2         2                           Chicken Bowl   

                                                           choice_description  \
0                                                                         nan   
1                                                                [Clementine]   
2                                                                     [Apple]   
3                                                                         nan   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   

   item_price  
0        2.39  
1        3.39  
2        3.39  
3        2.39  
4       16.98

In [119]:
# item_price is actually the total_price and quantity is useless!
chipotle_db.groupby('order_id')[['item_price']].sum()

item_price
order_id            
1              11.56
2              16.98
3              12.67
4              21.00
5              13.70
...              ...
1830           23.00
1831           12.90
1832           13.20
1833           23.50
1834           28.75

[1834 rows x 1 columns]

3. What are the most popular 3 items?

In [90]:
chipotle_db.groupby('item_name').quantity.sum().sort_values(ascending = False).head(3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

4. Which item has produced the most revenue?

In [92]:
chipotle_db.groupby('item_name').item_price.sum().sort_values(ascending = False).head(1)

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

5. Join the employees and titles DataFrames together.

In [98]:
employees_df.head()

emp_no  birth_date first_name last_name gender   hire_date
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4   10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12

In [99]:
titles_df.head()

emp_no            title   from_date     to_date
0   10001  Senior Engineer  1986-06-26  9999-01-01
1   10002            Staff  1996-08-03  9999-01-01
2   10003  Senior Engineer  1995-12-03  9999-01-01
3   10004         Engineer  1986-12-01  1995-12-01
4   10004  Senior Engineer  1995-12-01  9999-01-01

In [124]:
emp_titles = employees_df.merge(titles_df, left_on='emp_no', right_on='emp_no', how='inner', indicator=True)
emp_titles.head()

emp_no  birth_date first_name last_name gender   hire_date  \
0   10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1   10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2   10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4   10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   

             title   from_date     to_date _merge  
0  Senior Engineer  1986-06-26  9999-01-01   both  
1            Staff  1996-08-03  9999-01-01   both  
2  Senior Engineer  1995-12-03  9999-01-01   both  
3         Engineer  1986-12-01  1995-12-01   both  
4  Senior Engineer  1995-12-01  9999-01-01   both

6. For each title, find the hire date of the employee that was hired most recently with that title.

In [138]:
# filter out historical
# current_emp_titles = pd.read_sql('SELECT to_date = 9999-01-01 FROM titles', get_db_url('employees'))
emp_titles['current_title'] = emp_titles.to_date == '9999-01-01'
emp_titles

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date _merge  current_title  
0       Senior Engineer  1986-06-26  9999-01-01   both          False  
1                 Staff  1996-08-03  9999-01-01   both          False  
2       Senior Engineer  1995-12-03  9999-01-01   both          False  
3              Engineer  1986-12-01  1995-12-01   both          False  
4       Senior Engineer  1995-12-01  9999-01-01   both          False  
...                 ...         ...         ...    ...            ...  
443303         Engineer  1987-08-30  1992-08-29   both          False  
443304  Senior Engineer  1992-08-29  9999-01-01   both          False  
443305     Senior Staff  1998-12-27  9999-01-01   both          False  
443306            Staff  1993-12-27  1998-12-27   both          False  
443307         Engineer  1997-11-30  9999-01-01   both          False  

[443308 rows x 11 columns]

In [139]:
emp_titles.groupby('title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

### Extra Pandas Exercises and Resources

https://www.w3resource.com/python-exercises/pandas/index.php

https://towardsdatascience.com/20-pandas-functions-that-will-boost-your-data-analysis-process-f5dfdb2f9e05

https://github.com/guipsamora/pandas_exercises

https://github.com/ajcr/100-pandas-puzzles

### More Practice!

For even more practice with pandas, you can do the exercises from the SQL module, but instead of using SQL to do the aggregation, sorting, joining, etc, use pandas. That is, read the data from all of the tables into pandas dataframes and manipulate the dataframes.